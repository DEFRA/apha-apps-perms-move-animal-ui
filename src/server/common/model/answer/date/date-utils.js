/**
 * @typedef {{ day: string, month: string, year: string }} DateData
 */

export const MONTH_DAYS = 31
export const YEAR_MONTHS = 12

/** @param {Date} inputDate */
export const createDateAsUTC = (inputDate) =>
  new Date(
    Date.UTC(inputDate.getFullYear(), inputDate.getMonth(), inputDate.getDate())
  )

/** @param {DateData | undefined} date */
export const toJSDate = (date) => {
  return new Date(
    Date.UTC(Number(date?.year), Number(date?.month) - 1, Number(date?.day))
  )
}

/**
 * @param {DateData} inputDate
 * @returns {boolean}
 */
export const isFutureDate = (inputDate) => {
  const currentDate = createDateAsUTC(new Date())
  const date = toJSDate(inputDate)

  return date > currentDate
}
