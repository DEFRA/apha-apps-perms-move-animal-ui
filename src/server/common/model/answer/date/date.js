import Joi from 'joi'
import { AnswerModel } from '../answer-model.js'
import { validateAnswerAgainstSchema } from '../validation.js'
import { NotImplementedError } from '../../../helpers/not-implemented-error.js'

/** @import {AnswerViewModelOptions} from '../answer-model.js' */
/** @import {AnswerValidationResult} from '../validation.js' */

const missingSchema = ({ message }) =>
  Joi.string().trim().required().messages({
    'any.required': message,
    'string.empty': message
  })

/**
 * export @typedef {{
 *  hint: string,
 *  validation: {
 *    missingDate: { message: string },
 *    missingDay: { message: string },
 *    missingMonth: { message: string },
 *    missingYear: { message: string },
 *    invalidDay: { message: string },
 *    invalidMonth: { message: string },
 *    invalidYear: { message: string },
 *    invalidDate: { message: string },
 *    yearPattern: { message: string, pattern: RegExp },
 *    futureDate: { message: string }
 *  }
 * }} DateConfig
 */

/**
 * @typedef {{ day: string, month: string, year: string }} DateData
 */

/**
 * @typedef {AnswerValidationResult & {subfields: (keyof DateData)[]}} ValidationResultWithSubfields
 */

/**
 * @augments {AnswerModel<DateData>}
 */
export class DateAnswer extends AnswerModel {
  // eslint-disable-next-line jsdoc/require-returns-check
  /** @returns {DateConfig} */
  get config() {
    return /** @type {any} */ (this.constructor).config
  }

  // eslint-disable-next-line jsdoc/require-returns-check
  /** @returns {DateConfig} */
  static get config() {
    throw new NotImplementedError()
  }

  /**
   * @returns {DateData | undefined}
   */
  get value() {
    return this._data
  }

  get html() {
    const date = new Date(
      Number(this.value?.year),
      Number(this.value?.month) - 1,
      Number(this.value?.day)
    )
    return `${date.getDate()} ${date.toLocaleString('default', { month: 'long' })} ${date.getFullYear()}`
  }

  // eslint-disable-next-line @typescript-eslint/class-literal-property-style
  get template() {
    return 'model/answer/date/date.njk'
  }

  /**
   * @returns {DateData}
   */
  toState() {
    const day = this.value?.day
    const month = this.value?.month
    const year = this.value?.year

    return {
      day: day !== undefined ? day.trim() : '',
      month: month !== undefined ? month.trim() : '',
      year: year !== undefined ? year.trim() : ''
    }
  }

  /** @returns {ValidationResultWithSubfields} */
  validate() {
    const { validation } = this.config
    const missingDay = validateAnswerAgainstSchema(
      missingSchema(validation.missingDay),
      this.value?.day
    )
    const missingMonth = validateAnswerAgainstSchema(
      missingSchema(validation.missingMonth),
      this.value?.month
    )
    const missingYear = validateAnswerAgainstSchema(
      missingSchema(validation.missingYear),
      this.value?.year
    )

    const convertMessage = (subfield, { isValid, errors }) => ({
      isValid: isValid,
      errors: { [`date-${subfield}`]: errors.value },
      subfields: [subfield]
    })

    if (!missingDay.isValid && !missingMonth.isValid && !missingYear.isValid) {
      return {
        isValid: false,
        errors: { 'date-day': { text: validation.missingDate.message } },
        subfields: ['day', 'month', 'year']
      }
    } else if (!missingDay.isValid) {
      return convertMessage('day', missingDay)
    } else if (!missingMonth.isValid) {
      return convertMessage('month', missingMonth)
    } else if (!missingYear.isValid) {
      return convertMessage('year', missingYear)
    }

    const numberSchema = Joi.string().trim().regex(/^\d+$/)
    const validDaySchema = numberSchema.custom((value, helper) => {
      if (Number(value) < 1 || Number(value) > 31) {
        return helper.message({
          'object.any': ''
        })
      }
    })

    if (validDaySchema.validate(this.value?.day).error) {
      return {
        isValid: false,
        errors: { 'date-day': { text: validation.invalidDay.message } },
        subfields: ['day']
      }
    }

    const validMonthSchema = numberSchema.custom((value, helper) => {
      if (Number(value) < 1 || Number(value) > 12) {
        return helper.message({
          'object.any': ''
        })
      }
    })
    if (validMonthSchema.validate(this.value?.month).error) {
      return {
        isValid: false,
        errors: { 'date-month': { text: validation.invalidMonth.message } },
        subfields: ['month']
      }
    }

    if (numberSchema.validate(this.value?.year).error) {
      return {
        isValid: false,
        errors: { 'date-year': { text: validation.invalidYear.message } },
        subfields: ['year']
      }
    }

    if (
      Joi.string()
        .trim()
        .pattern(validation.yearPattern.pattern)
        .validate(this.value?.year).error
    ) {
      return {
        isValid: false,
        errors: { 'date-year': { text: validation.yearPattern.message } },
        subfields: ['year']
      }
    }

    const isValidDate = ({ year, month, day }) => {
      const date = new Date(Number(year), Number(month) - 1, Number(day))
      return (
        date.getFullYear() === Number(year) &&
        date.getMonth() === Number(month) - 1 &&
        date.getDate() === Number(day)
      )
    }

    if (this.value !== undefined && !isValidDate(this.value)) {
      return {
        isValid: false,
        errors: { 'date-day': { text: validation.invalidDate.message } },
        subfields: ['day', 'month', 'year']
      }
    }

    return {
      isValid: true,
      errors: {},
      subfields: []
    }
  }

  /**
   * @param {AnswerViewModelOptions} options
   */
  viewModel({ validate, question }) {
    const {
      isValid,
      errors = {},
      subfields = []
    } = validate ? this.validate() : { isValid: true }

    const classesWithErrors = (subfieldErroring, originalClasses) => {
      return subfieldErroring
        ? `${originalClasses} govuk-input--error`
        : originalClasses
    }

    const errorMessage = isValid
      ? {}
      : {
          text:
            errors['date-day']?.text ??
            errors['date-month']?.text ??
            errors['date-year']?.text
        }

    const viewModel = {
      fieldset: {
        legend: {
          text: question,
          classes: 'govuk-fieldset__legend--l',
          isPageHeading: true
        }
      },
      id: 'date',
      hint: {
        text: this.config.hint
      },
      items: [
        {
          classes: classesWithErrors(
            subfields.includes('day'),
            'govuk-input--width-2'
          ),
          name: 'day',
          value: this.value?.day
        },
        {
          classes: classesWithErrors(
            subfields.includes('month'),
            'govuk-input--width-2'
          ),
          name: 'month',
          value: this.value?.month
        },
        {
          classes: classesWithErrors(
            subfields.includes('year'),
            'govuk-input--width-4'
          ),
          name: 'year',
          value: this.value?.year
        }
      ]
    }

    if (validate) {
      viewModel.errorMessage = errorMessage
    }

    return viewModel
  }

  /**
   * @param {object} fields
   */
  _extractFields({ day, month, year }) {
    return /** @type {DateData} */ ({ day, month, year })
  }

  /**
   * @param {DateData| undefined} state
   * @returns {DateAnswer}
   */
  static fromState(state) {
    return new this(state)
  }
}
