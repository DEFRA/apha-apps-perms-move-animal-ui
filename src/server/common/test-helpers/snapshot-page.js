import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
