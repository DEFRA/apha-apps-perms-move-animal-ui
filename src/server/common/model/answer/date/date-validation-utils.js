import { toBSTDate } from './date-utils.js'

/**
 * @import { DateData } from './date-utils.js'
 * @import { ValidationResultWithSubfields } from './date.js'
 */

/** @param {string} str */
export const isDigits = (str) => str.match(/^\d*$/) !== null

/**
 * @param {string | undefined} value
 */
export const isMissing = (value) => !value

/** @param {string} str */
export const isFourDigits = (str) => str.match(/^\d{4}$/) !== null

/** @param {DateData} value */
export const isValidDate = (value) => {
  const date = toBSTDate(value)
  return (
    date.getFullYear() === Number(value.year) &&
    date.getMonth() === Number(value.month) - 1 &&
    date.getDate() === Number(value.day)
  )
}

/**
 * @param {(keyof DateData)[]} subfields
 * @param {string} message
 * @returns {ValidationResultWithSubfields}
 */
export const createError = (subfields, message) => ({
  isValid: false,
  errors: subfields.reduce((acc, subfield) => {
    acc[`date-${subfield}`] = { text: message }
    return acc
  }, {}),
  subfields
})

/**
 * @param {string} str
 * @param {number} min
 * @param {number} max
 */
export const isZeroPaddedDigitBetween = (str, min, max) => {
  const zeroPaddedRegex = /^((0?[1-9])|([1-9]\d))$/
  return (
    str.match(zeroPaddedRegex) !== null &&
    Number(str) >= min &&
    Number(str) <= max
  )
}

/**
 * @param {string} message
 * @returns {ValidationResultWithSubfields}
 */
export const allFieldsError = (message) => ({
  isValid: false,
  errors: { 'date-day': { text: message } },
  subfields: ['day', 'month', 'year']
})

/**
 * @param {keyof DateData} subfield
 * @param {string} message
 * @returns {ValidationResultWithSubfields}
 */
export const fieldError = (subfield, message) => ({
  isValid: false,
  errors: { [`date-${subfield}`]: { text: message } },
  subfields: [subfield]
})
