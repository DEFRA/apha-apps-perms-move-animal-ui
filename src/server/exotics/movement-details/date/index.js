import { QuestionPage } from '~/src/server/common/model/page/question-page-model.js'
import { ExoticsQuestionPageController } from '~/src/server/exotics/question-page-controller.js'
import { DateAnswer } from '~/src/server/common/model/answer/date/date.js'
import { checkAnswersPage } from '../check-answers/index.js'

/** @import { DateConfig } from '~/src/server/common/model/answer/date/date.js' */
/** @import { ServerRegisterPluginObject } from '@hapi/hapi' */

const questionKey = 'date'

export class Answer extends DateAnswer {
  /** @type { DateConfig } */
  static config = {
    hint: 'For example, 7 3 2025',
    validation: {
      missingDate: {
        message: 'Enter the expected movement date'
      },
      missingDay: {
        message: 'Expected movement date must include a day'
      },
      missingMonth: {
        message: 'Expected movement date must include a month'
      },
      missingYear: {
        message: 'Expected movement date must include a year'
      },
      invalidDay: {
        message: 'Expected movement day must be a real date'
      },
      invalidMonth: {
        message: 'Expected movement month must be a number between 1 and 12'
      },
      invalidYear: {
        message: 'Expected movement year must be a real date'
      },
      nonFourDigitYear: {
        message: 'Expected movement year must include 4 numbers'
      },
      invalidDate: {
        message: 'Expected movement date must be a real date'
      }
    }
  }
}

export class DatePage extends QuestionPage {
  urlPath = '/exotics/movement-details/date'

  questionKey = questionKey
  sectionKey = 'movementDetails'
  question = 'When do you expect the movement to happen?'

  Answer = Answer

  nextPage() {
    return checkAnswersPage
  }
}

export const datePage = new DatePage()

/** @satisfies {ServerRegisterPluginObject<void>} */
export const date = new ExoticsQuestionPageController(datePage).plugin()
