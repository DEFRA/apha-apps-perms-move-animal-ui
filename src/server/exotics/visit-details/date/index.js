import { QuestionPage } from '~/src/server/common/model/page/question-page-model.js'
import { ExoticsQuestionPageController } from '~/src/server/exotics/question-page-controller.js'
import { DateAnswer } from '~/src/server/common/model/answer/date/date.js'
import { checkAnswersPage } from '../check-answers/index.js'

/** @import { DateConfig } from '~/src/server/common/model/answer/date/date.js' */
/** @import { ServerRegisterPluginObject } from '@hapi/hapi' */

const questionKey = 'date'

export class Answer extends DateAnswer {
  /** @type { DateConfig } */
  static config = {
    hint: 'For example, 7 3 2025',
    validation: {
      missingDate: {
        message: 'Enter the expected visit date'
      },
      missingDay: {
        message: 'Expected visit date must include a day'
      },
      missingMonth: {
        message: 'Expected visit date must include a month'
      },
      missingYear: {
        message: 'Expected visit date must include a year'
      },
      invalidDay: {
        message: 'Expected visit day must be a real date'
      },
      invalidMonth: {
        message: 'Expected visit date must be a number between 1 and 12'
      },
      invalidYear: {
        message: 'Expected visit year must be a real date'
      },
      nonFourDigitYear: {
        message: 'Expected visit year must include 4 numbers'
      },
      invalidDate: {
        message: 'Expected visit date must be a real date'
      }
    }
  }
}

export class DatePage extends QuestionPage {
  urlPath = '/exotics/visit-details/date'

  questionKey = questionKey
  sectionKey = 'visitDetails'
  question = 'When do you expect the visit to happen?'

  Answer = Answer

  nextPage() {
    return checkAnswersPage
  }
}

export const datePage = new DatePage()

/** @satisfies {ServerRegisterPluginObject<void>} */
export const date = new ExoticsQuestionPageController(datePage).plugin()
