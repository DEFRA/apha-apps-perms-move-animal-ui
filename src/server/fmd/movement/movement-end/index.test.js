import { describePageSnapshot } from '~/src/server/common/test-helpers/snapshot-page.js'
import { Answer, movementEndPage } from './index.js'
import { DateAnswer } from '~/src/server/common/model/answer/date/date.js'
import { CheckAnswersPage } from '../check-answers/index.js'

const sectionKey = 'movement'
const questionKey = 'movementEnd'
const pageUrl = '/fmd/movement-details/end-date'
const page = movementEndPage
const question = 'What date does your movement end?'

const payload = {
  day: '01',
  month: '01',
  year: '2035'
}

describe('Answer', () => {
  it('should be a Date input', () => {
    expect(new Answer(payload)).toBeInstanceOf(DateAnswer)
  })

  describe('config validation options', () => {
    it('should have hint text', () => {
      expect(Answer.config.hint).toBe('For example, 7 3 2025')
    })

    it('should have missing date validation message', () => {
      expect(Answer.config.validation.missingDate).toEqual({
        message: 'Enter the movement end date'
      })
    })

    it('should have missing day validation message', () => {
      expect(Answer.config.validation.missingDay).toEqual({
        message: 'Movement end date must include a day'
      })
    })

    it('should have missing month validation message', () => {
      expect(Answer.config.validation.missingMonth).toEqual({
        message: 'Movement end date must include a month'
      })
    })

    it('should have missing year validation message', () => {
      expect(Answer.config.validation.missingYear).toEqual({
        message: 'Movement end date must include a year'
      })
    })

    it('should have invalid day validation message', () => {
      expect(Answer.config.validation.invalidDay).toEqual({
        message: 'Movement end day must be a real date'
      })
    })

    it('should have invalid month validation message', () => {
      expect(Answer.config.validation.invalidMonth).toEqual({
        message: 'Movement end month must be a number between 1 and 12'
      })
    })

    it('should have invalid year validation message', () => {
      expect(Answer.config.validation.invalidYear).toEqual({
        message: 'Movement end year must be a real date'
      })
    })

    it('should have non four digit year validation message', () => {
      expect(Answer.config.validation.nonFourDigitYear).toEqual({
        message: 'Movement end year must include 4 numbers'
      })
    })

    it('should have invalid date validation message', () => {
      expect(Answer.config.validation.invalidDate).toEqual({
        message: 'Movement end date must be a real date'
      })
    })

    it('should have past date validation message', () => {
      expect(Answer.config.validation.pastDate).toEqual({
        message: 'Movement end date must be in the future'
      })
    })

    it('should not have future date validation', () => {
      expect(Answer.config.validation.futureDate).toBeUndefined()
    })

    it('should not explicitly set isPageHeading (defaults to true)', () => {
      expect(Answer.config.isPageHeading).toBeUndefined()
    })
  })

  describe('validation', () => {
    it('should validate successfully with valid future date', () => {
      const answer = new Answer({ day: '15', month: '12', year: '2025' })
      const result = answer.validate()
      expect(result.isValid).toBe(true)
      expect(result.errors).toEqual({})
    })

    it('should fail validation when completely empty', () => {
      const answer = new Answer(undefined)
      const result = answer.validate()
      expect(result.isValid).toBe(false)
      expect(result.errors['date-day'].text).toBe('Enter the movement end date')
    })

    it('should fail validation when all fields are empty strings', () => {
      const answer = new Answer({ day: '', month: '', year: '' })
      const result = answer.validate()
      expect(result.isValid).toBe(false)
      expect(result.errors['date-day'].text).toBe('Enter the movement end date')
    })

    it('should fail validation when day is missing', () => {
      const answer = new Answer({ day: '', month: '12', year: '2025' })
      const result = answer.validate()
      expect(result.isValid).toBe(false)
      expect(result.errors['date-day'].text).toBe(
        'Movement end date must include a day'
      )
    })

    it('should fail validation when month is missing', () => {
      const answer = new Answer({ day: '15', month: '', year: '2025' })
      const result = answer.validate()
      expect(result.isValid).toBe(false)
      expect(result.errors['date-month'].text).toBe(
        'Movement end date must include a month'
      )
    })

    it('should fail validation when year is missing', () => {
      const answer = new Answer({ day: '15', month: '12', year: '' })
      const result = answer.validate()
      expect(result.isValid).toBe(false)
      expect(result.errors['date-year'].text).toBe(
        'Movement end date must include a year'
      )
    })

    it('should fail validation with invalid day (0)', () => {
      const answer = new Answer({ day: '0', month: '12', year: '2025' })
      const result = answer.validate()
      expect(result.isValid).toBe(false)
      expect(result.errors['date-day'].text).toBe(
        'Movement end day must be a real date'
      )
    })

    it('should fail validation with invalid day (32)', () => {
      const answer = new Answer({ day: '32', month: '12', year: '2025' })
      const result = answer.validate()
      expect(result.isValid).toBe(false)
      expect(result.errors['date-day'].text).toBe(
        'Movement end day must be a real date'
      )
    })

    it('should fail validation with invalid month (0)', () => {
      const answer = new Answer({ day: '15', month: '0', year: '2025' })
      const result = answer.validate()
      expect(result.isValid).toBe(false)
      expect(result.errors['date-month'].text).toBe(
        'Movement end month must be a number between 1 and 12'
      )
    })

    it('should fail validation with invalid month (13)', () => {
      const answer = new Answer({ day: '15', month: '13', year: '2025' })
      const result = answer.validate()
      expect(result.isValid).toBe(false)
      expect(result.errors['date-month'].text).toBe(
        'Movement end month must be a number between 1 and 12'
      )
    })

    it('should fail validation with non-numeric year', () => {
      const answer = new Answer({ day: '15', month: '12', year: 'abc' })
      const result = answer.validate()
      expect(result.isValid).toBe(false)
      expect(result.errors['date-year'].text).toBe(
        'Movement end year must be a real date'
      )
    })

    it('should fail validation with three digit year', () => {
      const answer = new Answer({ day: '15', month: '12', year: '202' })
      const result = answer.validate()
      expect(result.isValid).toBe(false)
      expect(result.errors['date-year'].text).toBe(
        'Movement end year must include 4 numbers'
      )
    })

    it('should fail validation with invalid date (29 Feb on non-leap year)', () => {
      const answer = new Answer({ day: '29', month: '2', year: '2025' })
      const result = answer.validate()
      expect(result.isValid).toBe(false)
      expect(result.errors['date-day'].text).toBe(
        'Movement end date must be a real date'
      )
    })

    it('should fail validation with past date', () => {
      const answer = new Answer({ day: '1', month: '1', year: '2020' })
      const result = answer.validate()
      expect(result.isValid).toBe(false)
      expect(result.errors['date-day'].text).toBe(
        'Movement end date must be in the future'
      )
    })

    it('should validate successfully with leap year date', () => {
      const answer = new Answer({ day: '29', month: '2', year: '2028' })
      const result = answer.validate()
      expect(result.isValid).toBe(true)
      expect(result.errors).toEqual({})
    })
  })
})

describe('MovementEndPage', () => {
  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct question', () => {
    expect(page.question).toBe(question)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(Answer)
  })

  describe('nextPage', () => {
    it('should return CheckAnswersPage for any value', () => {
      const nextPage = page.nextPage()
      expect(nextPage).toBeInstanceOf(CheckAnswersPage)
    })
  })

  describePageSnapshot({
    describes: 'content',
    it: 'should render expected response and content',
    pageUrl
  })
})
