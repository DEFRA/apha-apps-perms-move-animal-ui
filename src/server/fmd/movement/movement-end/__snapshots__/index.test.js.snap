// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovementEndPage content should render expected response and content 1`] = `
"
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset" role="group" aria-describedby="date-hint">
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
      <h1 class="govuk-fieldset__heading">
        What date does your movement end?
      </h1>
    </legend>
    <div id="date-hint" class="govuk-hint">
      For example, 7 3 2025
    </div>
    <div class="govuk-date-input" id="date">
      <div class="govuk-date-input__item">
        <div class="govuk-form-group">
          <label class="govuk-label govuk-date-input__label" for="date-day">
            Day
          </label>
          <input class="govuk-input govuk-date-input__input govuk-input--width-2" id="date-day" name="day" type="text" inputmode="numeric">
        </div>
      </div>
      <div class="govuk-date-input__item">
        <div class="govuk-form-group">
          <label class="govuk-label govuk-date-input__label" for="date-month">
            Month
          </label>
          <input class="govuk-input govuk-date-input__input govuk-input--width-2" id="date-month" name="month" type="text" inputmode="numeric">
        </div>
      </div>
      <div class="govuk-date-input__item">
        <div class="govuk-form-group">
          <label class="govuk-label govuk-date-input__label" for="date-year">
            Year
          </label>
          <input class="govuk-input govuk-date-input__input govuk-input--width-4" id="date-year" name="year" type="text" inputmode="numeric">
        </div>
      </div>
    </div>
  </fieldset>
</div>

            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
